### ISSUE TITLE
[Feature request] Gradient of cholesky_inverse, cholesky_solve

### ISSUE THREAD

**Issue by alexandrebone** [pytorch/pytorch#4669]
Hello everyone, 

Thank you for your amazing work. 
I would be interested by the implementation of the first derivative of the potri or/and potrs operators. I can use matrix inversion instead, but it comes with a performance drop cost.

Thank you again, 
Best, 

Alex

**Comment by SsnL** [comment]
I can work on this but it will be after finishing the tasks I have.

**Comment by el3ment** [comment]
Would also love this!

**Comment by mohakbhardwaj** [comment]
Hi, I would like to know if there was any further development on this feature as I am working on a project where I have to differentiate through potrs.
Thanks!

**Comment by el3ment** [comment]
@mohakbhardwaj I'm not sure if this will be helpful -- but the following code solves y = Ox for x* in a way that is differentiable

```
      # This method is slower than using the QR decomposition, but torch.qr does not yet have gradients defined
      # see: https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L618
      # to see if things have changed
      # Instead, we use the cholesky decomposition of O'O 
      U = torch.potrf(O.t().matmul(O), upper=False)
      rhs = Y.matmul(O)
      z = torch.trtrs(rhs.t(), U, transpose=False, upper=False)[0]
      x0 = torch.trtrs(z, U, transpose=True, upper=False)[0]
```

**Comment by mohakbhardwaj** [comment]
>       rhs = Y.matmul(O)

This method works when the input matrices are 2D, but does not work for batches. Did you have any luck with batches?

**Comment by el3ment** [comment]
It should work with batches - what version of pytorch are you using?

Robert

**Comment by mohakbhardwaj** [comment]
My version is 1.0.1.post2. I get the following error message when I try to run it for batches:
`RuntimeError: invalid argument 2: A should have 2 dimensions, but has 3 at /pytorch/aten/src/TH/generic/THTensorLapack.cpp:153`

**Comment by el3ment** [comment]
Which operation fails? Are these cpu or cuda tensors?

**Comment by mohakbhardwaj** [comment]
I think the failure is in torch.trtrs, the triangular solve operation. These are all cpu tensors

**Comment by el3ment** [comment]
Try cuda tensors.. i wonder if trtrs is batched for cuda, but not for cpu.

Robert

**Comment by mohakbhardwaj** [comment]
I tried with cuda tensors and got the same error as before.

**Comment by rgommers** [comment]
Note: in v1.2.0 `potri` and `potrs` were removed, replaced with `cholesky_inverse` and `cholesky_solve`. Those do have batched versions (see gh-7500), so perhaps that helps make your use case work @mohakbhardwaj

**Comment by calincru** [comment]
It's nice that we now have batched versions of those functions, but what about the gradients?  Is anyone working on this?

**Comment by rgommers** [comment]
@calincru no, that would still be this issue.

**Comment by fritzo** [comment]
I use the following workarounds for lack of batching in `cholesky_inverse` and lack of gradients in `cholesky_solve`. I'm guessing it should be easy to move this logic into PyTorch.
```py
def cholesky_solve(b, u):
    "Like :func:`torch.cholesky_solve` but supports gradients."
    if not b.requires_grad and not u.requires_grad:
        return b.cholesky_solve(u)
    x = b.triangular_solve(u, upper=False).solution
    return x.triangular_solve(u, upper=False, transpose=True).solution

def cholesky_inverse(u):
    "Like :func:`torch.cholesky_inverse` but supports batching and gradients."
    if u.dim() == 2:
        return u.cholesky_inverse()
    return cholesky_solve(torch.eye(u.size(-1)).expand(u.size()), u)
```

**Comment by vishwakftw** [comment]
The derivative of `cholesky_solve` is available here : https://github.com/pytorch/pytorch/pull/26185/

**Comment by vishwakftw** [comment]
The derivative of `cholesky_inverse` is available here : #26451
